blockchain/state_manager/src/lib.rs

```diff
@@ -223,10 +369,276 @@ where
+                })?
+                .next()
+                .await
+            {
+                match subscriber {
+                    HeadChange::Revert(_tipset) => {
+                        if candidate_tipset.is_some() {
+                            candidate_tipset = None;
+                            candidate_receipt = None;
+                        }
+                    }
+                    HeadChange::Apply(tipset) => {
+                        if candidate_tipset
+                            .as_ref()
+                            .map(|s| s.epoch() >= s.epoch() + tipset.epoch())
+                            .unwrap_or_default()
+                        {
+                            return Ok((candidate_tipset, candidate_receipt));
+                        }
+                        let poll_receiver = receiver.try_recv().map_err(|e| {
+                            Error::Other(format!("Could not receieve from channel {:?}", e))
+                        })?;
+                        if poll_receiver.is_some() {
+                            block_revert
+                                .write()
+                                .await
+                                .insert(tipset.key().to_owned(), true);
+                        }
+
+                        let message_var = (message.from(), &message.sequence());
+                        let maybe_receipt = Self::tipset_executed_message(
+                            &*block_store,
+                            &tipset,
+                            &cid,
+                            message_var,
+                        )?;
+                        if let Some(receipt) = maybe_receipt {
+                            if confidence == 0 {
+                                return Ok((Some(tipset), Some(receipt)));
+                            }
+                            candidate_tipset = Some(tipset);
+                            candidate_receipt = Some(receipt)
+                        }
+                    }
+                    _ => (),
+                }
+            }
+
+            Ok((None, None))
+        });
+
+        futures.push(subscriber_poll);
+        let search_back_poll = task::spawn(async move {
+            let (back_tipset, back_receipt) = task.await?;
+            let should_revert = *reverts
+                .read()
+                .await
+                .get(back_tipset.key())
+                .unwrap_or(&false);
+            let larger_height_of_head = height_of_head >= back_tipset.epoch() + confidence;
+            if !should_revert && larger_height_of_head {
+                return Ok((Some(Arc::new(back_tipset)), Some(back_receipt)));
+            }
+
+            Ok((None, None))
+        });
+        futures.push(search_back_poll);
+
+        futures.next().await.ok_or_else(|| Error::Other("wait_for_message could not be completed due to failure of subscriber poll or search_back functionality".to_string()))?
+        // Ok((None, None));
```

Comment:
You're losing the error message here for this static string, is there a reason it's dropped here?

Can you also remove the commented out code too?

Reply:
Yes I can. Oh no it is an Option on ok_or_else so there isn't an error. The error is returned as part of the futures.next() since futures.next().await returrns an option

Reply:
ohhhh my bad, I misread :)